### PR TITLE
Add an option to mark duplicates by read group.

### PR DESCRIPTION
**Pull request by whitwham**
This adds an option to take the read group into account when marking duplicates.

The stats have been changed to display duplicates by read groups after the totals.  There is also an option to output stats in JSON format.

### CODE REVIEW

**Comment by whitwham** [comment]
Even though they are valid JSON, I'm not entirely convinced the JSON stats are usefully formatted

**Comment by daviesrob** [comment]
Rebased and tidied up some odd strings that got into one of the commits.  Apologies for the forced push.

File: bam_markdup.c [hunk 1]
```diff
@@ -265,7 +289,7 @@ static int64_t calc_score(bam1_t *b)
    read is leftmost of the pair. */
 
 
-static int make_pair_key(md_param_t *param, key_data_t *key, bam1_t *bam, long *warnings) {
+static int make_pair_key(md_param_t *param, key_data_t *key, bam1_t *bam, uint8_t rg_num, long *warnings) {
```
Comment (daviesrob) [inline bam_markdup.c:292]: Why is this `uint8_t` and not `int`, as used elsewhere?  (Also in `make_single_key()`.)

Comment (whitwham) [reply bam_markdup.c:292]: I don't remember.  I think it was part of an earlier development phase.  I changed how it worked part way through.  I'll change it.

**Comment by daviesrob** [comment]
Thanks for the updates.  I've squashed in the fixes, and merged.

**Comment by whitwham** [comment]
Hopefully should have fixed the requests in #1143.

**Pull request merged**